Add `transform_func` and `ylim_func` parameters to `interact()`

gully opened the pull request:
Addresses Issue #578 

Outstanding issues:
- [ ] Transformations that change the number of cadences in a lightcurve are currently not supported and would result in either a silent fail or an uncaught bokeh error.
- [x] Add a test that passes in this kwarg
- [ ] Transformations currently must preserve the `cadenceno` attribute.

barentsen commented on lightkurve/interact.py:
```diff
@@ -517,6 +534,8 @@ def save_lightcurve():
                 selected_mask = np.isin(pixel_index_array, selected_indices)
                 lc_new = tpf.to_lightcurve(aperture_mask=selected_mask)
+                if transform_func is not None:
+                    lc_new = transform_func(lc_new)
```
It looks like lines 490-494, 534-538, and perhaps 438-440 do the same thing.  Would it be good to separate this into a single helper function (`_create_lc` or whatever) so we reduce the likelihood of bugs in future updates?

gully commented:
Now also addresses #577

gully commented:
I think this PR is ready for review/merge, yeah?

barentsen commented on lightkurve/tests/test_interact.py:
```diff
@@ -88,6 +88,22 @@ def test_custom_exported_filename():
+@pytest.mark.skipif(bad_optional_imports, reason="requires bokeh")
+def test_transform_and_ylim_funcs():
+    """Test the transform_func and ylim_func"""
+    import bokeh
```
I think this line can be removed:

```suggestion

```

barentsen commented on lightkurve/interact.py:
```diff
@@ -397,9 +403,30 @@ def show_interact_widget(tpf, notebook_url='localhost:8888',
+        Or it could be more complex, such as detrending the lightcurve.  In this
+        way, the interactive selection of aperture mask can be evaluated after
+        inspection of the transformed lightcurve.  See the tutorial for some
+        limitations. The transform_func is applied before saving a fits file.
```
I'd remove the sentence `See the tutorial for some limitations.` for now, unless you can link to a specific tutorial?

barentsen commented:
@gully This is very minor, but you may be able to improve the formatting of the example code in the parameters docstring slightly:
![Screenshot from 2019-10-08 09-44-34](https://user-images.githubusercontent.com/817669/66415560-b57e9b00-e9b0-11e9-9e5a-15564e47c792.png)

barentsen commented:
@gully Hmmm, `tpf.interact()` does not appear to work for me at all right now (both on this branch and on master, using both Chrome and Firefox).  Chrome reports the following error:
```
Couldn't process kernel message TypeError: Cannot read property 'trim' of undefined
    at output (eval at append_javascript (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:35081), <anonymous>:47:39)
    at Kernel._handle_output_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38184)
    at i (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:2)
    at Kernel._handle_iopub_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38211)
    at Kernel._finish_ws_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38003)
    at main.min.js?v=1bc4ae6ca8730320ff059b16de698691:37994
```

barentsen commented:
@gully `interact()` is working fine for me again, both on master and this branch, so you can ignore my previous post.

Can you have a quick look at the other comments I left above?  If you fix them we should be good to get this merged!

Finally, to enable an easy merge, can you rebase on master and edit `CHANGES.rst` to mention this new feature in the change log?  Thanks++.

barentsen commented:
Ping @gully -- do you want to look into the changes requested above so we can put this puppy to bed?  (Changes needed should be fairly small.)

gully commented:
OK, I think this PR is ready for merge once the tests pass.  The only slightly annoying thing is that you must preserve the number of cadences.  So for example, your transform function cannot include `lc.remove_outliers()`. Changing this requirement would be messy, since the cadence number mapping is hardcoded in several places.

gully commented:
Woohoo all tests have passed and the branch is rebased on master.  ping @barentsen Feel free to review/merge.

barentsen commented on CHANGES.rst:
```diff
@@ -8,7 +8,8 @@
 
-
+- Added user customization for lightcurve processing and plotting in the
+  `.interact()` method for TargetPixelFile objects. [#600]
```
Let's make this more descriptive:
```Added  ``transform_func`` and ``ylim_func`` keywords to ``TargetPixelFile.interact()`` to support custom light curve transformations and y-axis limits. [#600]```

barentsen commented on lightkurve/interact.py:
```diff
@@ -401,9 +406,28 @@ def show_interact_widget(tpf, notebook_url='localhost:8888',
+    transform_func: function
+        A function that transforms the lightcurve.  The function takes in a
+        LightCurve object as input and returns a LightCurve object as output.
+        For example, the function could be to normalize the lightcurve::
```
```suggestion
        For example, the function could be used to normalize the lightcurve::
```

barentsen commented:
@gully The example code in the keyword docstrings still doesn't seem to render correctly (see six comments up).  Can you fix?

gully commented:
Ahh, the docstring stems from the TargetPixelFile method, not the standalone interact.py sub-module.  Fixed the docs now, and added a nice looking Example section.

gully commented:
Weird.  The macOS test is failing because the design matrix class is attempting a `.plot()`, which appears to be looking for an OS X-specific backend gui.  We saw similar problems in the past, if I recall correctly.
